src/nni_manager/training_service/reusable/trialDispatcher.ts

```diff
@@ -0,0 +1,553 @@
+            this.isDeveloping = true;
+            this.runnerSettings.enableGpuCollector = true;
+        }
+    }
+
+    public async listTrialJobs(): Promise<TrialDetail[]> {
+        const trials: TrialDetail[] = [];
+
+        for (const key of this.trials.keys()) {
+            trials.push(await this.getTrialJob(key));
+        }
+
+        return trials;
+    }
+
+    public async getTrialJob(trialJobId: string): Promise<TrialDetail> {
+        const trial: TrialDetail | undefined = this.trials.get(trialJobId);
+        if (trial === undefined) {
+            throw new Error(`trial job ${trialJobId} not found`);
+        }
+
+        return trial;
+    }
+
+    public async submitTrialJob(form: TrialJobApplicationForm): Promise<TrialDetail> {
+        if (this.trialConfig === undefined) {
+            throw new Error(`trialConfig not initialized!`);
+        }
+
+        const trialId: string = uniqueString(5);
+
+        const environmentService = component.get<EnvironmentService>(EnvironmentService);
+        let trialWorkingFolder: string = "";
+        if (environmentService.hasStorageService) {
+            const storageService = component.get<StorageService>(StorageService);
+            trialWorkingFolder = storageService.joinPath('trials', trialId);
+        }
+        const trialJobDetail: TrialDetail = new TrialDetail(trialId, "WAITING", Date.now(), trialWorkingFolder, form);
+
+        this.trials.set(trialId, trialJobDetail);
+
+        return trialJobDetail;
+    }
+
+    // to support multi phase
+    public async updateTrialJob(trialJobId: string, form: TrialJobApplicationForm): Promise<TrialDetail> {
+        const trialDetail = await this.getTrialJob(trialJobId);
+        const environment = trialDetail.environment;
+        if (environment === undefined) {
+            throw new Error(`TrialDispatcher: trial ${trialJobId}'s env shouldn't be undefined in updateTrialJob.`);
+        }
+
+        const message = {
+            "trialId": trialJobId,
+            "parameters": form.hyperParameters,
+        }
+        await this.commandChannel.sendCommand(environment, SEND_TRIAL_JOB_PARAMETER, message);
+
+        return trialDetail;
+    }
+
+    public async cancelTrialJob(trialJobId: string, isEarlyStopped?: boolean | undefined): Promise<void> {
+        const trial = await this.getTrialJob(trialJobId);
+        switch (trial.status) {
+            case "RUNNING":
+            case "WAITING":
+            case "UNKNOWN":
+                {
+                    const environment = trial.environment;
+                    if (environment) {
+                        await this.commandChannel.sendCommand(environment, KILL_TRIAL_JOB, trial.id);
+                        trial.isEarlyStopped = isEarlyStopped;
+                        trial.status = trial.isEarlyStopped === true ?
+                            'EARLY_STOPPED' : 'USER_CANCELED';
```

Comment:
In current logic,  `trial.status` is not set here because the cancel could fail,   can we ensure the cancel succeed here?

Reply:
it cannot guarantee cancel succeed in any case, but it will try best... the runner will kill all child processes, which belows to trial code.